Add support for PACKET_FANOUT on Linux

tsnoam opened the pull request:
This work is based on PR #674 while taking into consideration the review
comments which caused the PR to be closed by the original author,
@xpahos.

mluscon commented:
Hi all, are there any issues with merging this PR?

Oppen commented:
I wonder if there's a more generic way to expose this, as many network stacks provide similar capabilities. Ones I definitely know of are DPDK and XDP (technically Linux, too, but you configure it by other means), but I expect most other mainstream *nixes to have equivalent features.

tsnoam commented:
@Oppen 
I get what you're saying that a consolidated way is for the best. 
However, since DPDK, XDP & PF_PACKET are so different in semantics, requiring the direct consumer to understand:
1. They're using one technology over another.
2. How the technology works (how to spawn threads, how to associate them with the kernel, who allocates the memory?).
3. As of "2" above, how to properly configure the fanout?
I would say that such consumer would want the full power to configure fanout without being masked out of control.

From what I gather, libpcap currently has initial support for DPDK, while it provides no support for XDP.
I suggest to currently implement fanout support for PF_PACKET as suggested in the PR (maybe even mark it as "unstable API" in the manpage). Once there will be enough code & knowledge about how to use other tech, it will be possible to either have separate APIs or find a way to consolidate them all.

Oppen commented:
@tsnoam I see your point, but I think masking implementation details is more or less the point of using libpcap as an abstraction. I'm OK-ish with marking it as an unstable API, although I have my doubts on that concept (eventually it has users and you can never drop it without facing their wrath).
Note though that I'm not a maintainer. My opinions and suggestions are just that, and carry no weight towards acceptance. Just in case you feel discouraged by my reluctance to make it platform-specific at the API level.

Also, it may be interesting to take a look at [this patchset](https://lists.openwall.net/netdev/2015/08/15/5).

Oppen commented on pcap-linux.c:
```diff
@@ -3730,6 +3730,21 @@ activate_sock(pcap_t *handle, int is_any_device)
+	 * Add the packet socket into FANOUT group, if needed.
+	 */
+#ifdef PACKET_FANOUT
+	if (handle->opt.fanout > 0) {
```
(Sorry this comment came in this late, I forgot to push the "start review" button weeks ago)

tsnoam replied:
good catch. long live the comments!
will be fixed.

tsnoam replied:
i decided to address this issue by adding an explicit `enable` flag to the function.

hacking around the `int` size is too risky.

Oppen replied:
I like that better, too.

Oppen commented on pcap-linux.c:
```diff
@@ -7136,6 +7151,28 @@ pcap_set_protocol_linux(pcap_t *p, int protocol)
+	return 0;
+#else
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+	    errno, "funout is not supported");
```
Typeo: "funout".

tsnoam replied:
oops. fixed.

Oppen replied:
It's funny that I made a typo pointing to a typo.

tsnoam commented:
> @tsnoam I see your point, but I think masking implementation details is more or less the point of using libpcap as an abstraction. I'm OK-ish with marking it as an unstable API, although I have my doubts on that concept (eventually it has users and you can never drop it without facing their wrath).
> Note though that I'm not a maintainer. My opinions and suggestions are just that, and carry no weight towards acceptance. Just in case you feel discouraged by my reluctance to make it platform-specific at the API level.
> 
> Also, it may be interesting to take a look at [this patchset](https://lists.openwall.net/netdev/2015/08/15/5).

1. Again, I get what you're saying about properly wrapping all various technologies into a single API, hiding implementation details from the user. However, waiting until everything was properly learned and consolidated may take too long and practically stops progress.
2. I looked at the patchset you've sent. It seems to be already merged into the kernel. If you were referring to the fact that there are more functions to wrap: well, this is exactly the point: you make small progress by slowly adding functions to competing APIs (packet socket vs dpdk vs ... ). One day you reach a point where you know enough of all technologies and feel comfortable wrapping them into a single API. But until then, why wait?

tsnoam commented:
Also, thanks for the review. I pushed and updated version of the code.

tsnoam commented on pcap/pcap.h:
```diff
@@ -352,6 +352,7 @@ PCAP_API const char *pcap_tstamp_type_val_to_description(int);
 
 #ifdef __linux__
 PCAP_API int	pcap_set_protocol_linux(pcap_t *, int);
+PCAP_API int	pcap_set_fanout_linux(pcap_t *, uint16_t, uint16_t);
```
1. You've commented on the previous version of this PR (not that it really matters, the diff is minor).
2. This line is inside `#ifdef __linux__` (just like `pcap_set_protocol_linux()`). It will only get built for Linux, where it's relevant.
Or, have I missed a problem you're pointing to?

Oppen replied:
1. Oh, sorry.
2. I see. I was aiming more at not breaking builds on other platforms even when used. But it's arguable that a system that expects fanout will have a design depending on it, so maybe breaking at build time is better. In any case, I think a nice message when the build breaks may be more user friendly than a link time missing symbol.

tsnoam replied:
* As far as I can see, the build of libpcap won't fail due to missing symbols.
* The failure will be witnessed during runtime of a client application calling `pcap_set_fanout_linux()` when `PACKET_FANOUT` wasn't defined. In such case, a descriptive error would be returned.
* The idea is that a function named `*_linux()` will be available on the Linux platform regardless if the inner implementation can be executed or not.
* Despite the above, I'm not "married" to this solution. We can hide `pcap_set_fanout_linux()` within `#ifdef PACKET_FANOUT`. I feel comfortable with that option as well.
* I have no strong opinion about both options.

Oppen replied:
1. I didn't mean libpcap's build, but a user's application.
2. Yes, that's OK. I'm thinking of, say, a FreeBSD user that mistakenly tries to build a program which depends on this (and, thus, on Linux), will the error *they* receive be clear enough?
3. I agree. It shouldn't *work* or be used on other platforms, but if it's clear to the user why the program doesn't work, then that's better.
4. I'm happier with the current solution than with `#ifdef`ing the whole function, mostly because it's easier to tell the users your program runs only on Linux than it is to tell them it runs on kernels above a given version, but might run in this and the other distros because they backport stuff and the like. However, as I said before, an application that depends on having fanout is probably an all or nothing situation, so my point here is rather moot.

Oppen replied:
Note using undeclared functions isn't itself an error (most compilers do issue warnings and allow turning them into errors, but that may be ignored), so a program written for Linux using this interface may build but fail to run.
Although now that I'm thinking more about that, I'm unsure there's a portable way to inform of that error to the downstream user/programmer.

tsnoam replied:
>     2\. Yes, that's OK. I'm thinking of, say, a FreeBSD user that mistakenly tries to build a program which depends on this (and, thus, on Linux), will the error _they_ receive be clear enough?

> Note using undeclared functions isn't itself an error (most compilers do issue warnings and allow turning them into errors, but that may be ignored), so a program written for Linux using this interface may build but fail to run.

Answer is similar:
1. Same behaviour like `pcap_set_protocol_linux` (another Linux only API which already exist in the code base).
2. Users using "unsafe" compiler flags (or more precisely, failing to add proper compiler flags)... I wish them best of luck with their code in so many ways...
3. I can start describing the various errors users may get on different points of the compilation/linkage/execution process in case the user decides to be creative in ignoring/working around compilation warnings, but I really do think it is meaningless (even the function has `linux` in its name, that should mean something, no?).

Oppen replied:
OK, that makes sense.

tsnoam commented:
> * I agree.
> 
> * That wasn't a counter argument or a suggestion for you to implement it now. It was just a "look at this cool thing", that may interest you because you're interested in something similar :)
>   I know it's merged, but the patch has a nice explanation about what it does, and I think it's a good way to point to features.

Oh, thanks for the reference then :-)

In any case, I feel that we do have an understanding that this PR is worthy of finding itself into upstream, with the open question of whether it should be marked as "unstable" pending maintainers decision.

Oppen commented:
> In any case, I feel that we do have an understanding that this PR is worthy of finding itself into upstream, with the open question of whether it should be marked as "unstable" pending maintainers decision.

I think it's definitely worthy of getting upstream. That doesn't mean improving it (whatever that means) is out of the question.

mcr commented on pcap_set_fanout_linux.3pcap:
```diff
@@ -0,0 +1,65 @@
+.SH DESCRIPTION
+On Linux network interface devices,
+.B pcap_set_fanout_linux()
+sets the fanout configuration to be used in the
```
Could you please explain what the affect of having packet fanout mode?
When would I turn it on and why?  Is there is a downside to it?  What if every application turned it on?

tsnoam replied:
In the AF_PACKET fanout mode, packet reception can be load balanced among processes/threads.

When "configuring" the socket to be used, you would be required to set a group_id, use the same group_id to achieve the required load balancing.

mcr replied:
Don't tell me. Put it in the manpage!

aouinizied commented:
@tsnoam I tried building libpcap from your branch. And I followed the doc. Everything compiles well and I test it using 3 parallel processes.

So I changed my pcap_open_live to the following in each process:

```c
pcap_handle = pcap_create((char*)pcap_file, err_open);
set = pcap_set_fanout_linux(pcap_handle, 1, 0, root_idx);
pcap_set_snaplen(pcap_handle, snaplen);
pcap_set_promisc(pcap_handle, promisc);
pcap_set_timeout(pcap_handle, 500);
pcap_activate(pcap_handle);
```

Which prints the desired output:

```bash
process (group) id: 0 fanout set: 0
process (group) id: 1 fanout set: 0
process (group) id: 2 fanout set: 0
```

However, when checking if FANOUT worked, I observe that each process still receives all packets. Note that I choose HASH to be flow-based (0 for mode value). Do you have any clue how to fix it?

xpahos commented:
@aouinizied it's very strange. I'm using the same construction in production:

```
#if defined(_linux_)
    int val = (getpid() & 0xffff) | (PACKET_FANOUT_LB << 16);
    if (setsockopt(pcap_fileno(pcap.Session()), SOL_PACKET, PACKET_FANOUT, &val, sizeof(val)) < 0) {
        Cerr << "Failed to set fanout: "
             << strerror(errno)
             << "\n";
        return;
    }
#endif
```

pcap_set_fanout_linux here do the same as 

```
int val = (getpid() & 0xffff) | (PACKET_FANOUT_LB << 16);
```

Does your root_idx is unique?

xpahos commented:
@aouinizied looks like I found the problem:

```
pcap_set_fanout_linux(pcap_t *p, int enable, uint16_t mode, uint16_t group_id)
```

I think it will be better to use:

```
set = pcap_set_fanout_linux(pcap_handle, 1, PACKET_FANOUT_LB, root_idx);
```

because the third parameter is **mode**.  I think it will be better to remove **enable** because it's not used in code.

aouinizied commented:
@xpahos Thanks for the feedback. I will replace PACKET_FANOUT_LB by 0 for FLOW_HASH mode, test, and get back to you.

aouinizied commented:
@xpahos checked that each root_idx is unique but the same observations.

```c
set = pcap_set_fanout_linux(pcap_handle, 1, 0, root_idx)
```

The only difference with your setup is the mode. You set it to PACKET_FANOUT_LB whereas I use [PACKET_FANOUT_HASH](https://elixir.bootlin.com/linux/v4.4/source/include/uapi/linux/if_packet.h#L60) as the third parameter.

xpahos commented:
@aouinizied yes, you right. PACKET_FANOUT_HASH should send packets only to one unique socket per each address port mapping.

aouinizied commented:
@xpahos another detail that may help reproducing is that I'm using:

```c
int rv_handle = pcap_next_ex(pcap_handle, &hdr, &data);
```

to poll the received packets. But this should not affect fanout right?

aouinizied commented:
@xpahos I found what was wrong. And it's clearly stated in the doc.
* group_id, as the name indicates, is a group of processes/threads.
* FANOUT will duplicate traffic across groups. (what I was observing)
* For a specific group_id. Each process/thread started with this group_id will receive 1/N of the traffic (where N is the number of processes/threads started with this group_id). 
* The mode defines load balancing strategy (on kernel side).

In my example replacing root_idx by 0 solve it and now I see traffic correctly load-balanced across my process. @tsnoam Thank you for this great contribution!


A detail that maybe need to be added to the doc is that for PACKET_FANOUT_HASH there is a known [**bug**](https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net-next.git/commit/?id=eb70db8756717b90c01ccc765fdefc4dd969fc74) on some Linux kernel versions. The bug consists in that a src->dst flow traffic will be hashed with another hash than dst->src and this will result in inconsistency for bidirectional flow-based apps.

tsnoam commented:
@aouinizied Sorry for the delay in response, I'm happy to see that you've eventually fixed the problem. Anyhow, most of the thanks are to @xpahos who had made the original PR, I have based on their work.

@xpahos that's a good catch regarding the `enable` flag. It appears that I have forgot to actually use it. I have just pushed a commit to do so.

tsnoam commented:
@mcr I've rebased my branch on top of the latest master and made appropriate fixes as of commit 97f59a7d035397c7b4f4889ef5657be811b97e7a

javedshakeel commented:
Hi. I wanted to try out fanout support in tcpdump. Is this mode controlled by a command line option to tcpdump or is it always enabled in this branch?

aouinizied commented:
@tsnoam is it possible to rebase your branch?

Many thanks for your contribution.

tsnoam commented:
@aouinizied 
I tried rebasing (and then build), and it passes cleanly.

have you encountered problems trying to rebase yourself?

aouinizied commented:
@tsnoam no but I'm using your fork (until it gets merged within libpcap master)  within my project (https://www.nfstream.org/docs/#building-nfstream-from-sources) and seems to be outdated. updating your branch with your rebase would be really helpful.

tsnoam commented:
Hi @aouinizied ,

As the branch can be directly merged to master, i mostly wait for the maintainers to pick it up.
Since rebasing will also require me to test that no internal changes dim it incompatible - things which can only be seen when running, I prefer to leave my PR and branch as it is, so the maintainers of of libpcap will be able to see on which commit I am based.

Personally, my systems work in production based on my branch, but I do not have the capacity at the moment to test a rebase over latest master.

tsnoam commented:
@mcr 
Do you have an estimation when this PR will be merged to master?
Is there something you expect me to change which I've missed?

Anything I can do to help, let me know!

infrastation commented:
Thank you for proposing these changes. If you consider them ready for merging, please rebase them on the current master branch and squash the commits together into one or more commits, so each change completely progress the working tree from one consistent working state to another.

tsnoam commented:
@mcr I've added the documentation as requested. I hope it is clear enough. If not, I'll be happy to fix as necessary.

@infrastation I do consider this change as ready for merging (FWIW, I've been using this code for 2+ years in production). As requested, I've squashed all commits and rebased it over current master. 

Regarding CI failure in AppVeyor, it seems unrelated to my change (`winflexbison` is missing from the system).

aouinizied commented:
@guyharris @tsnoam any plan to merge this PR?
I'm using this branch since more than 1 year on a production network and it works well. It will be good to have it as part of next official release.

Thanks,
Zied

aouinizied commented:
@guyharris any news regarding this PR?

aouinizied commented:
REMINDER: @infrastation @mcr any estimation of when this PR will be merged?

Zied

infrastation commented:
Thank you for waiting, there is quite a bit of backlog. One thing that does not look right in the proposed changes is that the fanout group membership is not related to anything else, so the new code is only a wrapper around `setsockopt()` on the `pcap_t.fd` on Linux. At the same time, on Linux `pcap_t.selectable_fd` equals `pcap_t.fd` and there is the existing **[pcap_get_selectable_fd](https://www.tcpdump.org/manpages/pcap_get_selectable_fd.3pcap.html)**(3PCAP) function, so at a glance it seems possible that the following would have the same or nearly the same effect without any changes to libpcap:
```c
int mode = mode << 16 | group_id & 0xffff;
int ret = setsockopt (
  pcap_get_selectable_fd (handle),
  SOL_PACKET,
  PACKET_FANOUT,
  &mode,
  sizeof (mode));
if (ret < 0) ...
```
Does it make sense?

tsnoam commented:
@infrastation this is a possible solution.

I do however wonder, if rejecting this PR will miss the point of libpcap being a holistic solution for packet capturing. So yes, some platforms and techniques have unique properties, but as long as their generic (pf_packet is a generic linux solution) I personally see it as something that I would want libpcap to do.

sfd commented:
I do think this would be useful to include in libpcap. Adding support might encourage other platforms to implement something similar, then a more generic API could be developed?
In either case the user application must be modified to make use of 'fanout', however a single pcap call would be more convenient than adding a (short) block of socket code to every application.
